File: app/services/feed_filters/pre_parse/html_entity_fixer.rb
require "nokogiri"

module FeedFilters
  module PreParse
    # 特定のRSSタグ内に含まれるHTMLエンティティを実際の文字に変換するフィルタ
    #
    # 一部のフィードでは<copyright>や<generator>などのタグに&copy;等のHTMLエンティティが
    # 含まれており、Feedjira/SAXMachineが正しく処理できず、以降の<title>や<description>を
    # パースできなくなる問題がある。
    #
    # このフィルタはタグ自体は残したまま、タグ内のHTMLエンティティのみを実際の文字に変換する。
    # これにより、フィルタの適用有無で「問題のあるフィード」を判定できる。
    class HtmlEntityFixer < Base
      # HTMLエンティティが問題を起こす可能性のあるタグ
      # これらのタグはアプリケーションで使用しないが、タグ自体は残す
      TARGET_TAGS = %w[
        copyright
        generator
      ].freeze

      def applicable?(xml_content, metadata = {})
        # 対象タグ内にHTMLエンティティが含まれている場合のみ適用
        TARGET_TAGS.any? do |tag|
          # まずタグが存在するかを単純な文字列検索で高速にチェック
          next false unless xml_content.include?("<#{tag}")

          # タグが存在する場合のみ、タグ内容を抽出してHTMLエンティティをチェック
          # 注: [^<]* を使うことで、巨大なファイルでのバックトラッキングを防止
          # copyright/generatorタグは通常テキストのみなのでこれで十分
          xml_content.match?(/<#{tag}[^>]*>([^<]*)&[a-zA-Z]+;/)
        end
      end

      def apply(xml_content, metadata = {})
        doc = Nokogiri::XML(xml_content)
        fixed_tags = []

        TARGET_TAGS.each do |tag|
          elements = doc.xpath("//channel/#{tag}")
          elements.each do |element|
            original_xml = element.to_xml
            # HTMLエンティティをデコードして実際の文字に変換
            # Nokogiri::HTML::DocumentFragmentを使ってHTMLとしてパースし、textで変換
            decoded_text = Nokogiri::HTML::DocumentFragment.parse(element.inner_html).text
            element.content = decoded_text

            # 変更があった場合のみ記録
            if element.to_xml != original_xml
              fixed_tags << tag unless fixed_tags.include?(tag)
            end
          end
        end

        if fixed_tags.any?
          mark_as_applied!(
            fixed_tags: fixed_tags,
            reason: "HTML entities in tags converted to actual characters"
          )
        end

        doc.to_xml
      end
    end
  end
end

File: fetcher/src/filters/html_entity_fixer.rs
//! copyright / generator の中の HTML エンティティを実際の文字に直す。
//! Ruby 版は Nokogiri で文書全体を読み直すが、ここでは対象タグの中身だけを書き換える。

use std::sync::LazyLock;

use regex::Regex;
use serde_json::json;

use super::PreParseFilter;

const TARGET_TAGS: [&str; 2] = ["copyright", "generator"];

static APPLICABLE: LazyLock<Vec<(&'static str, Regex)>> = LazyLock::new(|| {
    TARGET_TAGS
        .iter()
        .map(|tag| {
            (
                *tag,
                Regex::new(&format!(r"<{tag}[^>]*>([^<]*)&[a-zA-Z]+;")).unwrap(),
            )
        })
        .collect()
});

static ELEMENT: LazyLock<Vec<(&'static str, Regex)>> = LazyLock::new(|| {
    TARGET_TAGS
        .iter()
        .map(|tag| {
            (
                *tag,
                Regex::new(&format!(r"(<{tag}(?:\s[^>]*)?>)([^<]*)(</{tag}>)")).unwrap(),
            )
        })
        .collect()
});

pub struct HtmlEntityFixer;

fn xml_escape_text(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

impl PreParseFilter for HtmlEntityFixer {
    fn name(&self) -> &'static str {
        "HtmlEntityFixer"
    }

    fn apply(&self, xml: &str) -> Option<(String, serde_json::Value)> {
        let applicable = APPLICABLE
            .iter()
            .any(|(tag, re)| xml.contains(&format!("<{tag}")) && re.is_match(xml));
        if !applicable {
            return None;
        }

        let mut current = xml.to_string();
        let mut fixed_tags: Vec<&str> = Vec::new();
        for (tag, re) in ELEMENT.iter() {
            let replaced = re.replace_all(&current, |caps: &regex::Captures| {
                let decoded = html_escape::decode_html_entities(&caps[2]).to_string();
                format!("{}{}{}", &caps[1], xml_escape_text(&decoded), &caps[3])
            });
            if replaced != current {
                fixed_tags.push(tag);
                current = replaced.into_owned();
            }
        }

        if fixed_tags.is_empty() {
            return None;
        }
        Some((
            current,
            json!({ "fixed_tags": fixed_tags, "reason": "HTML entities in tags converted to actual characters" }),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::filters::PreParseFilter;

    #[test]
    fn decodes_entities_in_copyright_and_generator() {
        let xml = "<rss><channel><copyright>&copy; 2026</copyright><generator>G &trade;</generator><title>t &amp; u</title></channel></rss>";
        let (fixed, details) = HtmlEntityFixer.apply(xml).unwrap();
        assert!(fixed.contains("<copyright>© 2026</copyright>"));
        assert!(fixed.contains("<generator>G ™</generator>"));
        assert!(fixed.contains("<title>t &amp; u</title>"));
        assert_eq!(
            details["fixed_tags"],
            serde_json::json!(["copyright", "generator"])
        );
    }

    #[test]
    fn not_applicable_without_entities() {
        assert!(
            HtmlEntityFixer
                .apply("<rss><channel><copyright>2026</copyright></channel></rss>")
                .is_none()
        );
    }

    #[test]
    fn re_escapes_xml_special_chars() {
        let xml = "<rss><channel><copyright>&copy; A &lt; B</copyright></channel></rss>";
        let (fixed, _) = HtmlEntityFixer.apply(xml).unwrap();
        assert!(fixed.contains("<copyright>© A &lt; B</copyright>"));
    }
}
